polyply/src/backmap.py

```diff
@@ -11,15 +11,128 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
+import numpy as np
+import scipy.optimize
+import networkx as nx
 from .processor import Processor
 from .generate_templates import find_atoms
+from .linalg_functions import rotate_xyz
 """
 Processor implementing a template based back
 mapping to lower resolution coordinates for
 a meta molecule.
 """
 
+def find_edges(molecule, attr, value):
+    """
+    Find all edges of a `vermouth.molecule.Molecule` that have the
+    attribute `attr` with the corresponding value of `value`. Note
+    that this function is not order specific. That means (1, 2) is
+    the same as (2, 1).
+
+    Parameters
+    ----------
+    molecule: :class:vermouth.molecule.Molecule
+    attrs: tuple(str, str)
+         tuple of the attributes used in matching
+    values: tuple
+         corresponding values value
+
+    Returns
+    ----------
+    list
+       list of edges found
+    """
+    edges = []
+    for edge in molecule.edges:
+        node_a, node_b = edge
+
+        if attr[0] in molecule.nodes[node_a]:
+           if molecule.nodes[node_a][attr[0]] == value[0]:
+              if  attr[1] in molecule.nodes[node_b]:
+                 if molecule.nodes[node_b][attr[1]] == value[1]:
+                     edges.append(edge)
+
+        if attr[1] in molecule.nodes[node_a]:
+           if molecule.nodes[node_a][attr[1]] == value[1]:
+               if attr[0] in molecule.nodes[node_b]:
+                  if molecule.nodes[node_b][attr[0]] == value[0]:
+                      edges.append(edge)
+
+    return edges
+
+
+def orient_template(meta_molecule, current_node, template, built_nodes):
+
+    # 1. find neighbours at meta_mol level
+    neighbours = nx.all_neighbors(meta_molecule, current_node)
+
+    # 2. find connecting atoms at low-res level
+    edges = []
+    for resid in neighbours:
+        edge = find_edges(meta_molecule.molecule,
+                            ("resid", "resid"),
+                            (resid+1, current_node+1))
+        edges += edge
+
+    # 3. build coordinate system
+    ref_coords = np.zeros((3, len(edges)))
+    opt_coords = np.zeros((3, len(edges)))
+    for ndx, edge in enumerate(edges):
+        atom_a, atom_b = edge
+        resid_a = meta_molecule.molecule.nodes[atom_a]["resid"] - 1
+        resid_b = meta_molecule.molecule.nodes[atom_b]["resid"] - 1
+        if resid_a in built_nodes or resid_b in built_nodes:
+            if resid_a == current_node:
```

Comment:
These ifs need comments about the semantics.
A comment/docstring on the for loop/function about what's going on/the goal also won't hurt anyone